### PR TITLE
Add --security-opt flag for AppArmor security policy to invoker. Can …

### PR DESCRIPTION
**Pull request by jeremiaswerner**
Add --security-opt flag for AppArmor security policy to invoker. Can be configured using the INVOKER_CONTAINER_POLICY parameter in Env.sh.

### CODE REVIEW

**Comment by jeremiaswerner** [comment]
@perryibm This PR allows us to pass-through the apparmor policy for user containers dynamically per deployment. The policy itself is not part of this PR and will be delivered additionally. I've choosen to make the policy configurable because we want to have a specific policy for whisk instead of overriding `docker-default` policy, which allows us to be more restrictive on user containers.  I kept `docker-default` policy as the default for user containers if not overridden in the environment definition. However, `docker-default` will apply to the invoker itself.

**Comment by perryibm** [comment]
started PPG 558

**Comment by jeremiaswerner** [comment]
thanks for starting the PPG Perry

**Comment by perryibm** [comment]
Unfortunately, it seems to have failed repeatedly with the following in the invoker log:
```
Error response from daemon: Invalid --security-opt: "apparmor=docker-default"
```

**Comment by jeremiaswerner** [comment]
wow, it seems like the flag was introduce with docker 1.10.0
https://github.com/docker/docker/blob/master/CHANGELOG.md#1100-2016-02-04

I did not check and will flag the dependency

good catch @perryibm

**Comment by jeremiaswerner** [comment]
I need to further investigate why this failed. looking at the release log, the apparmor support should exists since docker 1.3.0
https://github.com/docker/docker/blob/master/CHANGELOG.md#130-2014-10-14

I'am confused with `--security-opts` (1.10.0) and `--security-opt` (1.3.0) now. so, I will retest with our current docker version and let you know

File: core/dispatcher/src/whisk/core/container/ContainerUtils.scala [hunk 1]
```diff
@@ -65,16 +65,17 @@ trait ContainerUtils extends Logging {
...
      */
-    def makeContainer(name: Option[String], image: String, network: String, env: Map[String, String], args: Array[String], limits: ActionLimits)(implicit transid: TransactionId): ContainerId = {
+    def makeContainer(name: Option[String], image: String, network: String, env: Map[String, String], args: Array[String], limits: ActionLimits, policy: Option[String])(implicit transid: TransactionId): ContainerId = {
         val nameOption = if (name isDefined) Array("--name", name.getOrElse("")) else Array.empty[String]
         val memoryArg = Array("-m", s"${limits.memory()}m")
         val capabilityArg = Array("--cap-drop", "NET_RAW", "--cap-drop", "NET_ADMIN")
         val consulServiceIgnore = Array("-e", "SERVICE_IGNORE=true")
         val fileHandleLimit = Array("--ulimit", "nofile=64:64")
         val processLimit = Array("--ulimit", "nproc=512:512")
+        val securityOpts = if (policy isDefined) Array("--security-opt", s"""apparmor:${policy}""" ) else Array.empty[String]
```
Comment (markusthoemmes) [inline core/dispatcher/src/whisk/core/container/ContainerUtils.scala]: `"""` not needed here

File: config/writePropertyFile.sh [hunk 2]
```diff
@@ -67,6 +67,11 @@ else
     echo "invoker.container.network=bridge" >> "$WHISK_HOME/whisk.properties"
 fi
 
+if [ "$INVOKER_CONTAINER_POLICY" != "" ]; then
+    echo "invoker.container.policy=$INVOKER_CONTAINER_POLICY" >> "$WHISK_HOME/whisk.properties"
+else
+    echo "invoker.container.policy=docker-default" >> "$WHISK_HOME/whisk.properties"
+fi
```
Comment (markusthoemmes) [inline config/writePropertyFile.sh]: I do believe, that with these in place `Map(invokerContainerPolicy -> "")` will never return `""` thus the `Option` will never be `None`. Intentional?

File: core/dispatcher/src/whisk/core/container/ContainerUtils.scala [hunk 3]
```diff
@@ -65,16 +65,17 @@ trait ContainerUtils extends Logging {
...
      */
-    def makeContainer(name: Option[String], image: String, network: String, env: Map[String, String], args: Array[String], limits: ActionLimits)(implicit transid: TransactionId): ContainerId = {
+    def makeContainer(name: Option[String], image: String, network: String, env: Map[String, String], args: Array[String], limits: ActionLimits, policy: Option[String])(implicit transid: TransactionId): ContainerId = {
         val nameOption = if (name isDefined) Array("--name", name.getOrElse("")) else Array.empty[String]
         val memoryArg = Array("-m", s"${limits.memory()}m")
         val capabilityArg = Array("--cap-drop", "NET_RAW", "--cap-drop", "NET_ADMIN")
         val consulServiceIgnore = Array("-e", "SERVICE_IGNORE=true")
         val fileHandleLimit = Array("--ulimit", "nofile=64:64")
         val processLimit = Array("--ulimit", "nproc=512:512")
+        val securityOpts = policy map { p => Array("--security-opt", s"apparmor:${p}" ) } getOrElse(Array.empty[String])
```
Comment (domdom82) [inline core/dispatcher/src/whisk/core/container/ContainerUtils.scala]: shouldn't this be `s"apparmor=${p}"` ?

**Comment by domdom82** [comment]
seems like in docker 1.9.1 the `--security-opt` option does not work if you put 
```
--security-opt apparmor=docker-default
```

you actually **have** to use the colon notation to make it work:
```
--security-opt apparmor:docker-default
```

Also since docker 1.11 the colon notation is deprecated (https://github.com/docker/docker/blob/master/CHANGELOG.md#1110-2016-04-13)
So this will need another follow-up change once we upgrade to docker >= 1.11


--security-opts is longer supported at all.

**Comment by domdom82** [comment]
rebased and squashed. removing wip tag.

**Comment by domdom82** [comment]
PG 279

**Comment by domdom82** [comment]
@perryibm pls review and merge / assign to committer if needed.

File: common/scala/src/main/scala/whisk/core/WhiskConfig.scala [hunk 4]
```diff
@@ -44,6 +44,7 @@ class WhiskConfig(
     val dockerImageTag = this(WhiskConfig.dockerImageTag)
 
     val invokerContainerNetwork = this(WhiskConfig.invokerContainerNetwork)
+    val invokerContainerPolicy = if (this(WhiskConfig.invokerContainerPolicy) == "") None else Some(this(WhiskConfig.invokerContainerPolicy))
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/WhiskConfig.scala]: We can probably get around by the default value of `""` and the `if..else` by using `Option.apply` like this:

```
val invokerContainerPolicy = Option(this(WhiskConfig.invokerContainerPolicy))
```

and setting the default value to `null` (gasp)

**Comment by perryibm** [comment]
You could but you still need to defend against someone setting it explicitly to the empty string which seems possible with configs.

**Comment by domdom82** [comment]
PG 279 was green. Assigning to @markusthoemmes for merge.

**Comment by markusthoemmes** [comment]
rebased and merged in 66a6447728e26e0daceb637dd4e909bcbfc81af5

**Pull request closed without merging**